##### Copyright 2019 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#@title MIT License
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

# Text classification with TensorFlow Hub: Movie reviews

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://www.tensorflow.org/tutorials/keras/text_classification_with_hub"><img src="https://www.tensorflow.org/images/tf_logo_32px.png" />View on TensorFlow.org</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View on GitHub</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/docs/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/download_logo_32px.png" />Download notebook</a>
  </td>
  <td>
    <a href="https://tfhub.dev/s?module-type=text-embedding"><img src="https://www.tensorflow.org/images/hub_logo_32px.png" />See TF Hub models</a>
  </td>
</table>

This notebook classifies movie reviews as *positive* or *negative* using the text of the review. This is an example of *binary*—or two-class—classification, an important and widely applicable kind of machine learning problem.

The tutorial demonstrates the basic application of transfer learning with [TensorFlow Hub](https://tfhub.dev) and Keras.

We'll use the [IMDB dataset](https://www.tensorflow.org/api_docs/python/tf/keras/datasets/imdb) that contains the text of 50,000 movie reviews from the [Internet Movie Database](https://www.imdb.com/). These are split into 25,000 reviews for training and 25,000 reviews for testing. The training and testing sets are *balanced*, meaning they contain an equal number of positive and negative reviews. 

This notebook uses [`tf.keras`](https://www.tensorflow.org/guide/keras), a high-level API to build and train models in TensorFlow, and [`tensorflow_hub`](https://www.tensorflow.org/hub), a library for loading trained models from [TFHub](https://tfhub.dev) in a single line of code. For a more advanced text classification tutorial using `tf.keras`, see the [MLCC Text Classification Guide](https://developers.google.com/machine-learning/guides/text-classification/).

In [3]:
!pip install -q tfds-nightly
!pip install -q tensorflow-hub

In [4]:
import os
import numpy as np

import tensorflow as tf
import tensorflow_hub as hub
import tensorflow_datasets as tfds

# Load compressed models from tensorflow_hub
os.environ["TFHUB_MODEL_LOAD_FORMAT"] = "COMPRESSED"

print("Version: ", tf.__version__)
print("Eager mode: ", tf.executing_eagerly())
print("Hub version: ", hub.__version__)
print("GPU is", "available" if tf.config.experimental.list_physical_devices("GPU") else "NOT AVAILABLE")

Version:  2.3.1
Eager mode:  True
Hub version:  0.9.0


GPU is available


## Download the IMDB dataset

The IMDB dataset is available on [imdb reviews](https://www.tensorflow.org/datasets/catalog/imdb_reviews) or on [TensorFlow datasets](https://www.tensorflow.org/datasets). The following code downloads the IMDB dataset to your machine (or the colab runtime):

In [5]:
# Split the training set into 60% and 40%, so we'll end up with 15,000 examples
# for training, 10,000 examples for validation and 25,000 examples for testing.
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews", 
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incompleteC476RF/imdb_reviews-train.tfrecord


Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incompleteC476RF/imdb_reviews-test.tfrecord


Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incompleteC476RF/imdb_reviews-unsupervised.tfrecord


Dataset imdb_reviews downloaded and prepared to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0. Subsequent calls will reuse this data.


## Explore the data 

Let's take a moment to understand the format of the data. Each example is a sentence representing the movie review and a corresponding label. The sentence is not preprocessed in any way. The label is an integer value of either 0 or 1, where 0 is a negative review, and 1 is a positive review.

Let's print first 10 examples.

In [6]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))
train_examples_batch

<tf.Tensor: shape=(10,), dtype=string, numpy=
array([b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it.",
       b'I have been known to fall asleep during films, but this is usually due to a combination of things including, really tired, being warm and comfortable on the sette and having just eaten a lot. However on this occasion I fell 

Let's also print the first 10 labels.

In [7]:
train_labels_batch

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0])>

## Build the model

The neural network is created by stacking layers—this requires three main architectural decisions:

* How to represent the text?
* How many layers to use in the model?
* How many *hidden units* to use for each layer?

In this example, the input data consists of sentences. The labels to predict are either 0 or 1.

One way to represent the text is to convert sentences into embeddings vectors. We can use a pre-trained text embedding as the first layer, which will have three advantages:

*   we don't have to worry about text preprocessing,
*   we can benefit from transfer learning,
*   the embedding has a fixed size, so it's simpler to process.

For this example we will use a **pre-trained text embedding model** from [TensorFlow Hub](https://tfhub.dev) called [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2).

There are many other pre-trained text embeddings from TFHub that can be used in this tutorial:

* [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2) - trained with the same NNLM architecture on the same data as [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2), but with a larger embedding dimension. Larger dimensional embeddings can improve on your task but it may take longer to train your model.
* [google/nnlm-en-dim128-with-normalization/2](https://tfhub.dev/google/nnlm-en-dim128-with-normalization/2) - the same as [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2), but with additional text normalization such as removing punctuation. This can help if the text in your task contains additional characters or punctuation.
* [google/universal-sentence-encoder/4](https://tfhub.dev/google/universal-sentence-encoder/4) - a much larger model yielding 512 dimensional embeddings trained with a deep averaging network (DAN) encoder.

And many more! Find more [text embedding models](https://tfhub.dev/s?module-type=text-embedding) on TFHub.

Let's first create a Keras layer that uses a TensorFlow Hub model to embed the sentences, and try it out on a couple of input examples. Note that no matter the length of the input text, the output shape of the embeddings is: `(num_examples, embedding_dimension)`.

In [8]:
embedding = "https://tfhub.dev/google/nnlm-en-dim50/2"
hub_layer = hub.KerasLayer(embedding, input_shape=[], 
                           dtype=tf.string, trainable=True)
hub_layer(train_examples_batch[:3])

<tf.Tensor: shape=(3, 50), dtype=float32, numpy=
array([[ 0.5423195 , -0.0119017 ,  0.06337538,  0.06862972, -0.16776837,
        -0.10581174,  0.16865303, -0.04998824, -0.31148055,  0.07910346,
         0.15442263,  0.01488662,  0.03930153,  0.19772711, -0.12215476,
        -0.04120981, -0.2704109 , -0.21922152,  0.26517662, -0.80739075,
         0.25833532, -0.3100421 ,  0.28683215,  0.1943387 , -0.29036492,
         0.03862849, -0.7844411 , -0.0479324 ,  0.4110299 , -0.36388892,
        -0.58034706,  0.30269456,  0.3630897 , -0.15227164, -0.44391504,
         0.19462997,  0.19528408,  0.05666234,  0.2890704 , -0.28468323,
        -0.00531206,  0.0571938 , -0.3201318 , -0.04418665, -0.08550783,
        -0.55847436, -0.23336391, -0.20782952, -0.03543064, -0.17533456],
       [ 0.56338924, -0.12339553, -0.10862679,  0.7753425 , -0.07667089,
        -0.15752277,  0.01872335, -0.08169781, -0.3521876 ,  0.4637341 ,
        -0.08492756,  0.07166859, -0.00670817,  0.12686075, -0.19326553,
 

Let's now build the full model:

In [9]:
model = tf.keras.Sequential()
model.add(hub_layer)
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
keras_layer (KerasLayer)     (None, 50)                48190600  
_________________________________________________________________
dense (Dense)                (None, 16)                816       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 17        
Total params: 48,191,433
Trainable params: 48,191,433
Non-trainable params: 0
_________________________________________________________________


The layers are stacked sequentially to build the classifier:

1. The first layer is a TensorFlow Hub layer. This layer uses a pre-trained Saved Model to map a sentence into its embedding vector. The pre-trained text embedding model that we are using ([google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2)) splits the sentence into tokens, embeds each token and then combines the embedding. The resulting dimensions are: `(num_examples, embedding_dimension)`. For this NNLM model, the `embedding_dimension` is 50.
2. This fixed-length output vector is piped through a fully-connected (`Dense`) layer with 16 hidden units.
3. The last layer is densely connected with a single output node.

Let's compile the model.

### Loss function and optimizer

A model needs a loss function and an optimizer for training. Since this is a binary classification problem and the model outputs logits (a single-unit layer with a linear activation), we'll use the `binary_crossentropy` loss function.

This isn't the only choice for a loss function, you could, for instance, choose `mean_squared_error`. But, generally, `binary_crossentropy` is better for dealing with probabilities—it measures the "distance" between probability distributions, or in our case, between the ground-truth distribution and the predictions.

Later, when we are exploring regression problems (say, to predict the price of a house), we will see how to use another loss function called mean squared error.

Now, configure the model to use an optimizer and a loss function:

In [10]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])

## Train the model

Train the model for 10 epochs in mini-batches of 512 samples. This is 10 iterations over all samples in the `x_train` and `y_train` tensors. While training, monitor the model's loss and accuracy on the 10,000 samples from the validation set:

In [11]:
history = model.fit(train_data.shuffle(10000).batch(512),
                    epochs=10,
                    validation_data=validation_data.batch(512),
                    verbose=1)

Epoch 1/10


 1/30 [>.............................] - ETA: 0s - loss: 0.7045 - accuracy: 0.5039

 3/30 [==>...........................] - ETA: 0s - loss: 0.7027 - accuracy: 0.5241

 4/30 [===>..........................] - ETA: 0s - loss: 0.7027 - accuracy: 0.5166

 6/30 [=====>........................] - ETA: 0s - loss: 0.6979 - accuracy: 0.5133

 8/30 [=======>......................] - ETA: 0s - loss: 0.6955 - accuracy: 0.5161

10/30 [=========>....................] - ETA: 0s - loss: 0.6947 - accuracy: 0.5105

12/30 [===========>..................] - ETA: 0s - loss: 0.6922 - accuracy: 0.5096

14/30 [=============>................] - ETA: 0s - loss: 0.6903 - accuracy: 0.5052

16/30 [===============>..............] - ETA: 0s - loss: 0.6879 - accuracy: 0.5052

18/30 [=================>............] - ETA: 0s - loss: 0.6862 - accuracy: 0.5040

20/30 [===================>..........] - ETA: 0s - loss: 0.6838 - accuracy: 0.5044

22/30 [=====================>........] - ETA: 0s - loss: 0.6813 - accuracy: 0.5051

24/30 [=======================>......] - ETA: 0s - loss: 0.6784 - accuracy: 0.5059

26/30 [=========================>....] - ETA: 0s - loss: 0.6756 - accuracy: 0.5082

28/30 [===========================>..] - ETA: 0s - loss: 0.6729 - accuracy: 0.5107

30/30 [==============================] - ETA: 0s - loss: 0.6706 - accuracy: 0.5131

30/30 [==============================] - 2s 59ms/step - loss: 0.6706 - accuracy: 0.5131 - val_loss: 0.6218 - val_accuracy: 0.5397


Epoch 2/10


 1/30 [>.............................] - ETA: 0s - loss: 0.6188 - accuracy: 0.5098

 3/30 [==>...........................] - ETA: 0s - loss: 0.6145 - accuracy: 0.5339

 5/30 [====>.........................] - ETA: 0s - loss: 0.6102 - accuracy: 0.5406

 7/30 [======>.......................] - ETA: 0s - loss: 0.6046 - accuracy: 0.5519

 9/30 [========>.....................] - ETA: 0s - loss: 0.6002 - accuracy: 0.5723

11/30 [==========>...................] - ETA: 0s - loss: 0.5956 - accuracy: 0.5872

13/30 [============>.................] - ETA: 0s - loss: 0.5917 - accuracy: 0.6007

15/30 [==============>...............] - ETA: 0s - loss: 0.5885 - accuracy: 0.6087

17/30 [================>.............] - ETA: 0s - loss: 0.5843 - accuracy: 0.6168

19/30 [==================>...........] - ETA: 0s - loss: 0.5809 - accuracy: 0.6238

21/30 [====================>.........] - ETA: 0s - loss: 0.5782 - accuracy: 0.6282

23/30 [======================>.......] - ETA: 0s - loss: 0.5735 - accuracy: 0.6362

25/30 [========================>.....] - ETA: 0s - loss: 0.5702 - accuracy: 0.6421

27/30 [==========================>...] - ETA: 0s - loss: 0.5656 - accuracy: 0.6482

29/30 [============================>.] - ETA: 0s - loss: 0.5613 - accuracy: 0.6548

30/30 [==============================] - 2s 53ms/step - loss: 0.5611 - accuracy: 0.6551 - val_loss: 0.5122 - val_accuracy: 0.7236


Epoch 3/10


 1/30 [>.............................] - ETA: 0s - loss: 0.4983 - accuracy: 0.7617

 3/30 [==>...........................] - ETA: 0s - loss: 0.4892 - accuracy: 0.7513

 5/30 [====>.........................] - ETA: 0s - loss: 0.4764 - accuracy: 0.7625

 7/30 [======>.......................] - ETA: 0s - loss: 0.4708 - accuracy: 0.7734

 9/30 [========>.....................] - ETA: 0s - loss: 0.4639 - accuracy: 0.7808

11/30 [==========>...................] - ETA: 0s - loss: 0.4588 - accuracy: 0.7855

13/30 [============>.................] - ETA: 0s - loss: 0.4556 - accuracy: 0.7856

15/30 [==============>...............] - ETA: 0s - loss: 0.4525 - accuracy: 0.7896

17/30 [================>.............] - ETA: 0s - loss: 0.4479 - accuracy: 0.7914

19/30 [==================>...........] - ETA: 0s - loss: 0.4459 - accuracy: 0.7949

21/30 [====================>.........] - ETA: 0s - loss: 0.4425 - accuracy: 0.7990

23/30 [======================>.......] - ETA: 0s - loss: 0.4390 - accuracy: 0.8007

25/30 [========================>.....] - ETA: 0s - loss: 0.4354 - accuracy: 0.8020

27/30 [==========================>...] - ETA: 0s - loss: 0.4324 - accuracy: 0.8030

29/30 [============================>.] - ETA: 0s - loss: 0.4290 - accuracy: 0.8053

30/30 [==============================] - 2s 53ms/step - loss: 0.4286 - accuracy: 0.8053 - val_loss: 0.4122 - val_accuracy: 0.8076


Epoch 4/10


 1/30 [>.............................] - ETA: 0s - loss: 0.3476 - accuracy: 0.8555

 3/30 [==>...........................] - ETA: 0s - loss: 0.3423 - accuracy: 0.8607

 5/30 [====>.........................] - ETA: 0s - loss: 0.3410 - accuracy: 0.8621

 7/30 [======>.......................] - ETA: 0s - loss: 0.3412 - accuracy: 0.8630

 9/30 [========>.....................] - ETA: 0s - loss: 0.3395 - accuracy: 0.8659

11/30 [==========>...................] - ETA: 0s - loss: 0.3356 - accuracy: 0.8665

13/30 [============>.................] - ETA: 0s - loss: 0.3325 - accuracy: 0.8672

15/30 [==============>...............] - ETA: 0s - loss: 0.3311 - accuracy: 0.8671

17/30 [================>.............] - ETA: 0s - loss: 0.3301 - accuracy: 0.8681

19/30 [==================>...........] - ETA: 0s - loss: 0.3266 - accuracy: 0.8710

21/30 [====================>.........] - ETA: 0s - loss: 0.3241 - accuracy: 0.8709

23/30 [======================>.......] - ETA: 0s - loss: 0.3215 - accuracy: 0.8725

25/30 [========================>.....] - ETA: 0s - loss: 0.3175 - accuracy: 0.8745

27/30 [==========================>...] - ETA: 0s - loss: 0.3155 - accuracy: 0.8752

29/30 [============================>.] - ETA: 0s - loss: 0.3134 - accuracy: 0.8761

30/30 [==============================] - 2s 53ms/step - loss: 0.3130 - accuracy: 0.8760 - val_loss: 0.3491 - val_accuracy: 0.8482


Epoch 5/10


 1/30 [>.............................] - ETA: 0s - loss: 0.2612 - accuracy: 0.8926

 3/30 [==>...........................] - ETA: 0s - loss: 0.2589 - accuracy: 0.9121

 5/30 [====>.........................] - ETA: 0s - loss: 0.2554 - accuracy: 0.9090

 7/30 [======>.......................] - ETA: 0s - loss: 0.2475 - accuracy: 0.9124

 9/30 [========>.....................] - ETA: 0s - loss: 0.2455 - accuracy: 0.9115

11/30 [==========>...................] - ETA: 0s - loss: 0.2443 - accuracy: 0.9110

13/30 [============>.................] - ETA: 0s - loss: 0.2403 - accuracy: 0.9121

15/30 [==============>...............] - ETA: 0s - loss: 0.2384 - accuracy: 0.9135

17/30 [================>.............] - ETA: 0s - loss: 0.2371 - accuracy: 0.9149

19/30 [==================>...........] - ETA: 0s - loss: 0.2342 - accuracy: 0.9164

21/30 [====================>.........] - ETA: 0s - loss: 0.2337 - accuracy: 0.9157

23/30 [======================>.......] - ETA: 0s - loss: 0.2331 - accuracy: 0.9149

25/30 [========================>.....] - ETA: 0s - loss: 0.2310 - accuracy: 0.9152

27/30 [==========================>...] - ETA: 0s - loss: 0.2296 - accuracy: 0.9162

29/30 [============================>.] - ETA: 0s - loss: 0.2291 - accuracy: 0.9165

30/30 [==============================] - 2s 54ms/step - loss: 0.2288 - accuracy: 0.9165 - val_loss: 0.3196 - val_accuracy: 0.8582


Epoch 6/10


 1/30 [>.............................] - ETA: 0s - loss: 0.2061 - accuracy: 0.9180

 3/30 [==>...........................] - ETA: 0s - loss: 0.1871 - accuracy: 0.9316

 5/30 [====>.........................] - ETA: 0s - loss: 0.1833 - accuracy: 0.9359

 7/30 [======>.......................] - ETA: 0s - loss: 0.1793 - accuracy: 0.9372

 9/30 [========>.....................] - ETA: 0s - loss: 0.1797 - accuracy: 0.9377

11/30 [==========>...................] - ETA: 0s - loss: 0.1784 - accuracy: 0.9387

13/30 [============>.................] - ETA: 0s - loss: 0.1762 - accuracy: 0.9401

15/30 [==============>...............] - ETA: 0s - loss: 0.1741 - accuracy: 0.9408

17/30 [================>.............] - ETA: 0s - loss: 0.1724 - accuracy: 0.9407

19/30 [==================>...........] - ETA: 0s - loss: 0.1726 - accuracy: 0.9412

21/30 [====================>.........] - ETA: 0s - loss: 0.1729 - accuracy: 0.9415

23/30 [======================>.......] - ETA: 0s - loss: 0.1725 - accuracy: 0.9421

25/30 [========================>.....] - ETA: 0s - loss: 0.1696 - accuracy: 0.9430

27/30 [==========================>...] - ETA: 0s - loss: 0.1698 - accuracy: 0.9423

29/30 [============================>.] - ETA: 0s - loss: 0.1685 - accuracy: 0.9434

30/30 [==============================] - 2s 54ms/step - loss: 0.1680 - accuracy: 0.9437 - val_loss: 0.3054 - val_accuracy: 0.8692


Epoch 7/10


 1/30 [>.............................] - ETA: 0s - loss: 0.1338 - accuracy: 0.9648

 3/30 [==>...........................] - ETA: 0s - loss: 0.1380 - accuracy: 0.9583

 5/30 [====>.........................] - ETA: 0s - loss: 0.1342 - accuracy: 0.9625

 7/30 [======>.......................] - ETA: 0s - loss: 0.1315 - accuracy: 0.9643

 9/30 [========>.....................] - ETA: 0s - loss: 0.1336 - accuracy: 0.9622

11/30 [==========>...................] - ETA: 0s - loss: 0.1343 - accuracy: 0.9606

13/30 [============>.................] - ETA: 0s - loss: 0.1352 - accuracy: 0.9587

15/30 [==============>...............] - ETA: 0s - loss: 0.1344 - accuracy: 0.9590

17/30 [================>.............] - ETA: 0s - loss: 0.1322 - accuracy: 0.9605

19/30 [==================>...........] - ETA: 0s - loss: 0.1307 - accuracy: 0.9604

21/30 [====================>.........] - ETA: 0s - loss: 0.1288 - accuracy: 0.9619

23/30 [======================>.......] - ETA: 0s - loss: 0.1259 - accuracy: 0.9628

25/30 [========================>.....] - ETA: 0s - loss: 0.1250 - accuracy: 0.9628

27/30 [==========================>...] - ETA: 0s - loss: 0.1244 - accuracy: 0.9626

29/30 [============================>.] - ETA: 0s - loss: 0.1230 - accuracy: 0.9629

30/30 [==============================] - 2s 53ms/step - loss: 0.1229 - accuracy: 0.9629 - val_loss: 0.3056 - val_accuracy: 0.8739


Epoch 8/10


 1/30 [>.............................] - ETA: 0s - loss: 0.0982 - accuracy: 0.9727

 3/30 [==>...........................] - ETA: 0s - loss: 0.0980 - accuracy: 0.9779

 5/30 [====>.........................] - ETA: 0s - loss: 0.1003 - accuracy: 0.9750

 7/30 [======>.......................] - ETA: 0s - loss: 0.0944 - accuracy: 0.9771

 9/30 [========>.....................] - ETA: 0s - loss: 0.0949 - accuracy: 0.9755

11/30 [==========>...................] - ETA: 0s - loss: 0.0936 - accuracy: 0.9757

13/30 [============>.................] - ETA: 0s - loss: 0.0917 - accuracy: 0.9766

15/30 [==============>...............] - ETA: 0s - loss: 0.0919 - accuracy: 0.9770

17/30 [================>.............] - ETA: 0s - loss: 0.0917 - accuracy: 0.9771

19/30 [==================>...........] - ETA: 0s - loss: 0.0913 - accuracy: 0.9770

21/30 [====================>.........] - ETA: 0s - loss: 0.0926 - accuracy: 0.9757

23/30 [======================>.......] - ETA: 0s - loss: 0.0925 - accuracy: 0.9761

25/30 [========================>.....] - ETA: 0s - loss: 0.0917 - accuracy: 0.9763

27/30 [==========================>...] - ETA: 0s - loss: 0.0905 - accuracy: 0.9766

29/30 [============================>.] - ETA: 0s - loss: 0.0897 - accuracy: 0.9772

30/30 [==============================] - 2s 53ms/step - loss: 0.0897 - accuracy: 0.9772 - val_loss: 0.3076 - val_accuracy: 0.8710


Epoch 9/10


 1/30 [>.............................] - ETA: 0s - loss: 0.0710 - accuracy: 0.9863

 3/30 [==>...........................] - ETA: 0s - loss: 0.0706 - accuracy: 0.9850

 5/30 [====>.........................] - ETA: 0s - loss: 0.0718 - accuracy: 0.9855

 7/30 [======>.......................] - ETA: 0s - loss: 0.0745 - accuracy: 0.9847

 9/30 [========>.....................] - ETA: 0s - loss: 0.0718 - accuracy: 0.9844

11/30 [==========>...................] - ETA: 0s - loss: 0.0697 - accuracy: 0.9844

13/30 [============>.................] - ETA: 0s - loss: 0.0681 - accuracy: 0.9848

15/30 [==============>...............] - ETA: 0s - loss: 0.0684 - accuracy: 0.9844

17/30 [================>.............] - ETA: 0s - loss: 0.0681 - accuracy: 0.9851

19/30 [==================>...........] - ETA: 0s - loss: 0.0670 - accuracy: 0.9856

21/30 [====================>.........] - ETA: 0s - loss: 0.0666 - accuracy: 0.9859

23/30 [======================>.......] - ETA: 0s - loss: 0.0661 - accuracy: 0.9857

25/30 [========================>.....] - ETA: 0s - loss: 0.0656 - accuracy: 0.9856

27/30 [==========================>...] - ETA: 0s - loss: 0.0663 - accuracy: 0.9852

29/30 [============================>.] - ETA: 0s - loss: 0.0653 - accuracy: 0.9856

30/30 [==============================] - 2s 53ms/step - loss: 0.0653 - accuracy: 0.9855 - val_loss: 0.3181 - val_accuracy: 0.8730


Epoch 10/10


 1/30 [>.............................] - ETA: 0s - loss: 0.0456 - accuracy: 0.9922

 3/30 [==>...........................] - ETA: 0s - loss: 0.0498 - accuracy: 0.9909

 5/30 [====>.........................] - ETA: 0s - loss: 0.0490 - accuracy: 0.9922

 7/30 [======>.......................] - ETA: 0s - loss: 0.0481 - accuracy: 0.9927

 9/30 [========>.....................] - ETA: 0s - loss: 0.0485 - accuracy: 0.9922

11/30 [==========>...................] - ETA: 0s - loss: 0.0484 - accuracy: 0.9924

13/30 [============>.................] - ETA: 0s - loss: 0.0474 - accuracy: 0.9928

15/30 [==============>...............] - ETA: 0s - loss: 0.0473 - accuracy: 0.9930

17/30 [================>.............] - ETA: 0s - loss: 0.0471 - accuracy: 0.9933

19/30 [==================>...........] - ETA: 0s - loss: 0.0470 - accuracy: 0.9927

21/30 [====================>.........] - ETA: 0s - loss: 0.0472 - accuracy: 0.9925

23/30 [======================>.......] - ETA: 0s - loss: 0.0471 - accuracy: 0.9924

25/30 [========================>.....] - ETA: 0s - loss: 0.0474 - accuracy: 0.9923

27/30 [==========================>...] - ETA: 0s - loss: 0.0474 - accuracy: 0.9925

29/30 [============================>.] - ETA: 0s - loss: 0.0470 - accuracy: 0.9925

30/30 [==============================] - 2s 53ms/step - loss: 0.0470 - accuracy: 0.9926 - val_loss: 0.3263 - val_accuracy: 0.8718


## Evaluate the model

And let's see how the model performs. Two values will be returned. Loss (a number which represents our error, lower values are better), and accuracy.

In [12]:
results = model.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model.metrics_names, results):
  print("%s: %.3f" % (name, value))

49/49 - 1s - loss: 0.3548 - accuracy: 0.8549


loss: 0.355
accuracy: 0.855


This fairly naive approach achieves an accuracy of about 87%. With more advanced approaches, the model should get closer to 95%.

## Further reading

* For a more general way to work with string inputs and for a more detailed analysis of the progress of accuracy and loss during training, see the [Text classification with preprocessed text](./text_classification.ipynb) tutorial.
* Try out more [text-related tutorials](https://www.tensorflow.org/hub/tutorials#text-related-tutorials) using trained models from TFHub.